### PR TITLE
avoid circular dependency usage between mcore and modelopt

### PR DESCRIPTION
**Pull request by dimapihtar**
- [x] I, the PR author, have personally reviewed every line of this PR.

# What does this PR do?

:warning: For major changes (either in lines of code or in its impact), please make sure to first share a design doc with the team. If you're unsure what's the best way to do so, contact @NVIDIA/mcore-oncall.

## Issue tracking

For PRs from open-source community contributors:

- **New features**: a linked issue is **required**. Please open a [feature request](https://github.com/NVIDIA/Megatron-LM/issues/new?template=feature_request.md) and reference it here before submitting the PR.
- **Small updates (bug fixes, minor improvements)**: a linked issue is **recommended** and will accelerate the PR review process.

Linked issue: 

### Pre-checks

- [ ] I have added relevant unit tests
- [ ] I have added relevant functional tests
- [ ] I have added proper typing to my code [Typing guidelines](https://docs.python.org/3/library/typing.html)
- [ ] I have added relevant documentation
- [ ] I have run the [autoformatter.sh](https://github.com/NVIDIA/Megatron-LM/blob/main/tools/autoformat.sh) on my PR

### Code review

Feel free to message or comment @NVIDIA/mcore-oncall to help accelerate your merge into main. The less complex your PR is, the faster it will be approved and merged!

All PRs start as **draft**. If you open a non-draft PR, it will be automatically converted to draft.

#### Step 1: Mark PR as "Ready for Review"

1. When your PR is ready, click **Ready for Review**.
2. An oncall reviewer is auto-assigned and expert reviewers are notified based on your changes.
   - Some PRs may jump straight to step 2. This is determined by `.github/CODEOWNERS`.

:warning: Only mark as ready once merge-conflicts are resolved and the CI is passing.
Final Review might get declined if these requirements are not fulfilled.

#### Step 2: Final Review

For PRs that change `megatron/core`, once all expert reviewers have approved, the `Final Review` label is applied **automatically** and final reviewers are assigned.

For PRs outside `megatron/core`, this step is skipped.

#### Step 3: Approved

Once all required reviewers have approved, the `Approved` label is applied **automatically**.

### Merge

Any member of [mcore-engineers](https://github.com/orgs/NVIDIA/teams/mcore-engineers) will be able to merge your PR.

### CODE REVIEW

File: megatron/training/checkpointing.py
Comment (AAnoosheh) [inline megatron/training/checkpointing.py:1]: Why don't these go into the other file too?

Comment (dimapihtar) [reply megatron/training/checkpointing.py:1]: there are only two places where we use these modules: `training/checkpointing.py` & `post_training/checkpointing.py` so I just put them here.

Do you think it'd be better to have separate file for these modelopt modules?

Comment (kevalmorabia97) [reply megatron/training/checkpointing.py:1]: We might need to use these functions in megatron bridge as well so having in megatron.core makes more sense

File: megatron/core/dist_checkpointing/strategies/modelopt.py [hunk 1]
```diff
@@ -0,0 +1,187 @@
...
+from ..validation import StrictHandling
+from .torch import TorchDistLoadShardedStrategy
+
+try:
+    import modelopt
+    import modelopt.torch.opt as mto
+    import modelopt.torch.utils.distributed as dist
+
+    has_nvidia_modelopt = True
+except (ModuleNotFoundError, ImportError):
```
Comment (kevalmorabia97) [inline megatron/core/dist_checkpointing/strategies/modelopt.py]: ModuleNotFoundError is subclass of ImportError
```suggestion
except ImportError:
```

File: megatron/core/dist_checkpointing/strategies/modelopt.py [hunk 2]
```diff
@@ -0,0 +1,187 @@
...
+from ..serialization import load, load_common_state_dict, save
+from ..validation import StrictHandling
+from .torch import TorchDistLoadShardedStrategy
+
+try:
+    import modelopt
+    import modelopt.torch.opt as mto
+    import modelopt.torch.utils.distributed as dist
+
+    has_nvidia_modelopt = True
```
Comment (kevalmorabia97) [inline megatron/core/dist_checkpointing/strategies/modelopt.py]: Where is has_nvidia_modelopt used?

**Review by kevalmorabia97 — APPROVED** [review]
LGTM

**Review by AAnoosheh — CHANGES_REQUESTED** [review]
IMPORTANT: Don't merge as we have further inquiries to the consequences of this change. (See Slack thread)

File: megatron/core/dist_checkpointing/strategies/modelopt.py (hunk 2 above)
Comment (dimapihtar) [reply megatron/core/dist_checkpointing/strategies/modelopt.py]: just a default style of guarding in mcore

**Review by AAnoosheh — APPROVED** [review]

**Review by cuichenx — APPROVED** [review]

**Review by kvareddy — APPROVED** [review]

**Review by jaredcasper — APPROVED** [review]
